fix: handle invalid `-rpcbind` port earlier

tdb3 opened the pull request:
Previously, when an invalid port was specified in `-rpcbind`, the `SplitHostPort()` return value in `HTTPBindAddresses()` was ignored and attempt would be made to bind to the default rpcbind port (with the host/port treated as a host).

This rearranges port checking code in `AppInitMain()` to handle the invalid port before reaching `HTTPBindAddresses()`.  Also adds a check in `HTTPBindAddresses()` as a defensive measure for future changes.

Adds then updates associated functional tests as well.

tdb3 commented:
Existing behavior:
```
src/bitcoind -rpcallowip=127.0.0.1 -rpcbind=127.0.0.1:18000 -rpcbind=[::1]:18001 -rpcbind=127.0.0.1:notaport -rpcbind=127.0.0.1:-18002 -rpcbind=[::1]:-18003
...
2024-08-20T02:13:49Z Command-line arg: rpcbind="127.0.0.1:18000"
2024-08-20T02:13:49Z Command-line arg: rpcbind="[::1]:18001"
2024-08-20T02:13:49Z Command-line arg: rpcbind="127.0.0.1:notaport"
2024-08-20T02:13:49Z Command-line arg: rpcbind="127.0.0.1:-18002"
2024-08-20T02:13:49Z Command-line arg: rpcbind="[::1]:-18003"
2024-08-20T02:13:49Z Using at most 125 automatic connections (1024 file descriptors available)
2024-08-20T02:13:49Z scheduler thread start
2024-08-20T02:13:49Z Binding RPC on address 127.0.0.1 port 18000
2024-08-20T02:13:49Z Binding RPC on address ::1 port 18001
2024-08-20T02:13:49Z Binding RPC on address 127.0.0.1:notaport port 18443
2024-08-20T02:13:49Z [libevent:warning] getaddrinfo: nodename nor servname provided, or not known
2024-08-20T02:13:49Z Binding RPC on address 127.0.0.1:notaport port 18443 failed.
2024-08-20T02:13:49Z Binding RPC on address 127.0.0.1:-18002 port 18443
2024-08-20T02:13:49Z [libevent:warning] getaddrinfo: nodename nor servname provided, or not known
2024-08-20T02:13:49Z Binding RPC on address 127.0.0.1:-18002 port 18443 failed.
2024-08-20T02:13:49Z Binding RPC on address [::1]:-18003 port 18443
2024-08-20T02:13:49Z [libevent:warning] getaddrinfo: nodename nor servname provided, or not known
2024-08-20T02:13:49Z Binding RPC on address [::1]:-18003 port 18443 failed.
2024-08-20T02:13:49Z Using random cookie authentication.
2024-08-20T02:13:49Z Generated RPC authentication cookie /home/dev/.bitcoin/regtest/.cookie
2024-08-20T02:13:49Z Permissions used for cookie: rw-------
2024-08-20T02:13:49Z Starting HTTP server with 4 worker threads
2024-08-20T02:13:49Z Using wallet directory /home/dev/.bitcoin/regtest/wallets
2024-08-20T02:13:49Z init message: Verifying wallet(s)…
2024-08-20T02:13:49Z Using /16 prefix for IP bucketing
2024-08-20T02:13:49Z init message: Loading P2P addresses…
2024-08-20T02:13:49Z Loaded 0 addresses from peers.dat  0ms
2024-08-20T02:13:49Z init message: Loading banlist…
2024-08-20T02:13:49Z SetNetworkActive: true
2024-08-20T02:13:49Z [error] Invalid port specified in -rpcbind: '127.0.0.1:notaport'
Error: Invalid port specified in -rpcbind: '127.0.0.1:notaport'
```

New behavior:
```
src/bitcoind -rpcallowip=127.0.0.1 -rpcbind=127.0.0.1:18000 -rpcbind=[::1]:18001 -rpcbind=127.0.0.1:notaport -rpcbind=127.0.0.1:-18002 -rpcbind=[::1]:-18003
...
2024-08-20T02:13:18Z Command-line arg: rpcallowip="127.0.0.1"
2024-08-20T02:13:18Z Command-line arg: rpcbind="127.0.0.1:18000"
2024-08-20T02:13:18Z Command-line arg: rpcbind="[::1]:18001"
2024-08-20T02:13:18Z Command-line arg: rpcbind="127.0.0.1:notaport"
2024-08-20T02:13:18Z Command-line arg: rpcbind="127.0.0.1:-18002"
2024-08-20T02:13:18Z Command-line arg: rpcbind="[::1]:-18003"
2024-08-20T02:13:18Z Using at most 125 automatic connections (1024 file descriptors available)
2024-08-20T02:13:18Z scheduler thread start
2024-08-20T02:13:18Z [error] Invalid port specified in -rpcbind: '127.0.0.1:notaport'
2024-08-20T02:13:18Z Unable to bind any endpoint for RPC server
2024-08-20T02:13:18Z [error] Unable to start HTTP server. See debug log for details.
```

ryanofsky commented:
Code review 9f19d10922d615c1e313eb7348fcc536642ad2e4.

I'm not sure this fix is ideal. For example if the GUI is used I believe this will change the error message from:

- Error: Invalid port specified in -rpcbind: '127.0.0.1:notaport'

to just:

- Unable to start HTTP server. See debug log for details.

I think it would be good to keep the new `return false` and error log in `HTTPBindAddresses`. But a more direct fix for this issue might be to just move the code that checks port numbers on line 1304 above the code that starts the http server on line 1219 in `AppInitMain`:

https://github.com/bitcoin/bitcoin/blob/d79ea809d28197b1b4e3748aa1715272b53601d0/src/init.cpp#L1304-L1316

https://github.com/bitcoin/bitcoin/blob/d79ea809d28197b1b4e3748aa1715272b53601d0/src/init.cpp#L1219-L1223

tdb3 committed:
test: add tests for invalid rpcbind ports

tdb3 commented:
> I think it would be good to keep the new `return false` and error log in `HTTPBindAddresses`. But a more direct fix for this issue might be to just move the code that checks port numbers on line 1304 above the code that starts the http server on line 1219 in `AppInitMain`:

Thank you for taking a lot.  Agreed, rearranging in `AppInitMain` seems to be the better approach.  Updated.  Also rearranged the commits so that the tests are introduced in the first commit and adjusted with the introduction of the fix.

ryanofsky commented on src/init.cpp:
```diff
@@ -1211,6 +1211,50 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     RegisterZMQRPCCommands(tableRPC);
 #endif
 
+    // Check port numbers
```
In commit "fix: handle invalid rpcbind port earlier" (8892150c2d8ea9e09046e288cecd4c35472c0267)

Not important, but to make AppInitMain function shorter and make the bugfix commit fix more readable you could consider adding an earlier refactoring commit the moves these for loops into a `CheckHostPortOptions` function, and have AppinitMain call it like `if (!CheckHostPortOptions(args)) return false;`

tdb3 replied:
Thanks, good suggestion.  Makes things cleaner.  Added.
Was also thinking about updating `ParseUint16()` to `ToIntegral()`, but that could be left for a follow-up, since we would also want to ensure adequate tests for `-port` and `-rpcport` with that update.

ryanofsky reviewed (approved):
Code review ACK 8892150c2d8ea9e09046e288cecd4c35472c0267. Looks good! Nice to have earlier feedback and clearer checking for these options. I left a suggestion, but it's not important, so feel free to ignore

tdb3 committed:
refactor: move host/port checking

Reduces the size of AppInitMain() by moving
checks to CheckHostPortOptions()

Co-Authored-By: Ryan Ofsky <ryan@ofsky.org>

tdb3 committed:
fix: handle invalid rpcbind port earlier

Previously, when an invalid port was specified
in `-rpcbind`, the `SplitHostPort()` return value
in `HTTPBindAddresses()` was ignored and attempt
would be made to bind to the default rpcbind port
(with the host/port treated as a host).

This rearranges port checking code in
`AppInitMain()` to handle the invalid
port before reaching `HTTPBindAddresses()`.

Also adjusts associated functional tests.

ryanofsky commented on src/httpserver.cpp:
```diff
@@ -374,7 +377,11 @@ static bool HTTPBindAddresses(struct evhttp* http)
             uint16_t port{http_port};
             std::string host;
-            SplitHostPort(strRPCBind, port, host);
+            if (!SplitHostPort(strRPCBind, port, host)) {
```
Code review ACK 51ba0e61273bec02833329ee6ac8fa219679cefa

(Not important, but another idea to make this commit smaller and clearer: I think it would be good to move the change in `src/httpserver.cpp` to a separate followup commit, because while it's a good change that improves code quality and error reporting, it is not really a part of the bugfix, and is more tangentially related to it.)

tdb3 replied:
Thanks.  Yes, that's correct (movement of CheckHostPortOptions() handles the bug before HTTPBindAddresses()). It's cleaner to break this out into its own commit. Updated.

ryanofsky reviewed (approved):
Code review ACK 51ba0e61273bec02833329ee6ac8fa219679cefa. Since last review, just split up the change and moved some code so this should be a little easier to review now.

I had another idea and left a suggestion but it is not important so please feel free to ignore.

tdb3 committed:
fix: increase rpcbind check robustness

Adds invalid rpcbind port checking to
`HTTPBindAddresses()`. While movement of
`CheckHostPortOptions()` in the previous
commit handles rcpbind port errors, updating
`HTTPBindAddresses()` port checking adds
a defensive measure for potential future
changes.

ryanofsky reviewed (approved):
Code review ACK 590d2fb0d8548d6d55fd1c738c8bcbe71a57aaf2

Just split up the last commit since the previous review to separate bugfix from cleanup

zaidmstrr commented:
Hey, I'm reviewing this PR, and I saw that the test `run_invalid_bind_test` you implemented in `rpc_bind.py` is not invoking without explicitly providing the `--ipv4` or `--ipv6` flags, so is there any particular reason for that?

tdb3 commented:
> Hey, I'm reviewing this PR, and I saw that the test `run_invalid_bind_test` you implemented in `rpc_bind.py` is not invoking without explicitly providing the `--ipv4` or `--ipv6` flags, so is there any particular reason for that?

Thank you for the review!
In `test/functional/test_runner.py`, `rpc_bind.py` is run with three variations (`--ipv4`, `--ipv6`, and `--nonloopback`), so there is coverage for parsing rpcbind with both IPv4 and IPv6 addresses/port binds (loopback).

zaidmstrr reviewed (commented):
ACK [590d2fb](https://github.com/bitcoin/bitcoin/pull/30679/commits/590d2fb0d8548d6d55fd1c738c8bcbe71a57aaf2)
I have reviewed and tested by manual and dry running the code to detect any edge cases and coding related erros. I also performed manual fuzzing by passing arguments in the CLI using bitcoind to check for any unexpected behaviours and to check whether the new code giving the updated results without any bypass or failure. 

tdb3 commented:
Rebased to address conflict

ryanofsky reviewed (approved):
Code review ACK e6994efe08b282dd9e46602bcbad69567fe91dcd

No changes since last review, just rebased due to conflict

zaidmstrr reviewed (approved):
Code review ACK [e6994ef](https://github.com/bitcoin/bitcoin/commit/e6994efe08b282dd9e46602bcbad69567fe91dcd)
All the things seem correct as before.

achow101 commented:
ACK e6994efe08b282dd9e46602bcbad69567fe91dcd
